### PR TITLE
buildah: new package

### PR DESCRIPTION
**Pull request by upsj**
Adds the buildah package, which provides a lightweight rootless framework for building container images based on Dockerfile/Containerfile specs.

### CODE REVIEW

File: var/spack/repos/builtin/packages/buildah/package.py
Comment (upsj) [inline var/spack/repos/builtin/packages/buildah/package.py:9]: The reason I made this a MakefilePackage is that it autodetects some btrfs dependencies (which should not be necessary for most builds) in the Makefile, but not in the regular Go build.

File: var/spack/repos/builtin/packages/buildah/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/buildah/package.py]: Why use the commit and not a sha256?

**Review by tldahlgren — DISMISSED** [review]
Confirmed the homepage and commit.

**Comment by tldahlgren** [comment]
Added @alecbcs since he has more `go` experience.

File: var/spack/repos/builtin/packages/buildah/package.py
Comment (upsj) [reply var/spack/repos/builtin/packages/buildah/package.py]: If you're asking, then it's likely missing familiarity with best practices ;)

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/buildah/package.py]: Not required, just curious since *most* people use sha256s.

That said, the commit is still considered a trusted download (https://spack.readthedocs.io/en/latest/packaging_guide.html#trusted-downloads).

Comment (upsj) [reply var/spack/repos/builtin/packages/buildah/package.py]: And I just learned that the archive hash stability issues in GitHub have been resolved, I only heard about the initial issues, not the fix for them

File: var/spack/repos/builtin/packages/buildah/package.py
Comment (alecbcs) [reply var/spack/repos/builtin/packages/buildah/package.py:9]: Do you know how they're doing the autodetection of btrfs dependencies in the makefile? It'd be better to have this as a `GoPackage` so that you get the same level of build isolation as in other Spack builds. Otherwise this is going to install dependencies into `$HOME/go` which isn't what most users want.

**Review by alecbcs — CHANGES_REQUESTED** [review]
One quick suggestion to fix the lower bound on the allowable version of `go` and one question about the MakefilePackage decision. If possible it'd be better to turn this into a GoPackage and still perform the same discovery as the Makefile.

File: var/spack/repos/builtin/packages/buildah/package.py
Comment (alecbcs) [inline var/spack/repos/builtin/packages/buildah/package.py]:
```suggestion
    depends_on("go@1.21:", type="build")
```

Requires `go >= 1.21.0` according to the [`go.mod` file in the repository.](https://github.com/containers/buildah/blob/v1.37.5/go.mod)

File: var/spack/repos/builtin/packages/buildah/package.py
Comment (upsj) [reply var/spack/repos/builtin/packages/buildah/package.py:9]: They check for the existence of the `btrfs/version.h` C++ header in the default search path, and set the `-tags` flag accordingly. Does `GoPackage` have an equivalent feature to set these tags? This would be the `btrfs_noversion exclude_graphdriver_btrfs` tags, probably together with `seccomp exclude_graphdriver_devicemapper` and optionally `libsubid systemd`

File: var/spack/repos/builtin/packages/buildah/package.py
Comment (alecbcs) [reply var/spack/repos/builtin/packages/buildah/package.py]: There's really not a fix for archive hash instabilities on GitHub. My preferred way of adding versions to new packages is via git commits these days. (Although I'd add both the commit and tag here which speeds up cloning.)

Comment (alecbcs) [inline var/spack/repos/builtin/packages/buildah/package.py]:
```suggestion
    version("1.37.5", tag="v1.37.5", commit="5fd40b989860984a00f6fc1539ff53caceca1325")
```

Suggestion to add the `v1.37.5` tag. This allows Spack to more quickly clone the ref from GitHub.

**Review by bernhardkaindl — COMMENTED** [review]
@upsj: I would like to set this to draft until the conversation with Alec about GoPackage is resolved:

This avoids running the GitLab pipeline until it is set to ready for review again.

## Notes:

### Discussion on GoPackage

The PR cannot be merged before the discussion about using GoPackage has been set to resolved (either by you or by Alec preferably)

### This PR is in state "Changes Requested"

Because Alec has set the flag for "_changed requested_" in his review, Alec has to re-review the PR with an approval before merging it can be unlocked.

----
Build results on Ubuntu 22.04.4 LTS: These changed specs were found `gh pr diff`:
- `buildah@1.37.5`

The following specs were built (or installed) to check this PR:
### Using linux-ubuntu22.04-skylake / gcc@13.1.0:
<details><summary>Click on this line to unfold and refold the dependencies of this build


```py
buildah@1.37.5 build_system=makefile
```
</summary>

Dependency tree (disabled variants are removed for the make the list the dependencies brief):
```py
buildah@1.37.5 build_system=makefile
    bats@1.10.0 build_system=generic
    bzip2@1.0.8+shared build_system=generic
        diffutils@3.10
    gcc-runtime@13.1.0 build_system=generic
    git@2.45.2+man+nls+perl+subtree
        autoconf@2.72
        automake@1.16.5
        curl@7.81.0+gssapi+ldap+libidn2+nghttp2 libs=shared,static tls=openssl
        expat@2.6.3+libbsd
            libbsd@0.12.2
                libmd@1.0.4
        gettext@0.22.5+bzip2+curses+git+libxml2+pic+shared+tar+xz
            libxml2@2.10.3+pic+shared
            ncurses@6.5+termlib abi=none patches=7a351bc
            tar@1.34 zip=pigz
                pigz@2.8 build_system=makefile
                zstd@1.5.6+programs build_system=makefile compression=none libs=shared,static
            xz@5.4.6 libs=shared,static
        libiconv@1.17 libs=shared,static
        libidn2@2.3.7
            libunistring@1.2
        libtool@2.4.7
            findutils@4.9.0 patches=440b954
        m4@1.4.19+sigsegv patches=9dc5fbd,bfdffa7
            libsigsegv@2.14
        openssh@9.8p1+gssapi
            krb5@1.21.2+shared
                bison@3.8.2
            libedit@3.1-20240808
            libxcrypt@4.4.35_api patches=4885da3
        openssl@3.3.1+shared build_system=generic certs=mozilla
            ca-certificates-mozilla@2023-05-30 build_system=generic
        pcre2@10.44+multibyte
        perl@5.40.0+cpanm+opcode+open+shared+threads build_system=generic
            berkeley-db@18.1.40+cxx+stl patches=26090f4,b231fcc
            gdbm@1.23
        zlib-ng@2.2.1+compat_strategies+opt+pic+shared
    glibc@2.35
    gmake@4.4.1 build_system=generic
    go@1.23.1 build_system=generic
        bash@5.2 patches=094b4fd,155853b,1c48cec,2500a3f,2fb107c,325c268,3882772,3ef9246,45cc5e1,4641dd4,5138f48,6a090cd,6b4bd92,78b5230,831b5f2,8334b88,96ee1f5,9715344,af90550,b6fc252,c7705e0,d1e0566,ece0eb5,ef73905,f42f2fe,f95a817
            readline@8.2 patches=bbf97f1
        go-bootstrap@1.20.6 build_system=generic
        grep@3.11
        sed@4.9
    go-md2man@2.0.4 build_system=go
    gpgme@1.20.0
        gnupg@2.4.5
            libgcrypt@1.11.0 patches=4e308ba
            libksba@1.6.7
            npth@1.7
            pinentry@1.3.1 gui=tty
        libgpg-error@1.50 patches=0b2a0ff
            gawk@5.3.1
                gmp@6.3.0+cxx libs=shared,static
                mpfr@4.2.1 libs=shared,static
                    autoconf-archive@2023.02.20
                    texinfo@7.1
    libassuan@2.5.7
    libseccomp@2.5.4+python
        gperf@3.1 patches=3dd36db
        py-cython@3.0.11 build_system=python_pip
            py-pip@23.1.2 build_system=generic
            py-setuptools@63.4.3 build_system=generic
            py-wheel@0.41.2 build_system=generic
            python@3.8.19+bz2+crypt+ctypes+dbm+libxml2+lzma+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl+uuid+zlib build_system=generic patches=0d98e93,4c24573,ebdca64,f2fd060
                libffi@3.4.6
                sqlite@3.46.0+column_metadata+dynamic_extensions+fts+rtree
                util-linux-uuid@2.40.2
            python-venv@1.0 build_system=generic
    pkgconf@2.2.0
    runc@1.1.13 build_system=makefile
```
</details>

Generated by https://github.com/spack/build-quality-tools/blob/main/build_pr_changes.py -a

**Comment by alecbcs** [comment]
@bernhardkaindl please don't accept suggestions. We should let the contributor accept / merge suggestions as a learning experience and to maintain control over their own branch.

File: var/spack/repos/builtin/packages/buildah/package.py
Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/buildah/package.py:9]: @alecbcs, do you have an answer for @upsj here?

**Comment by upsj** [comment]
@alecbcs @bernhardkaindl I tried getting the spack-built `buildah` to work as well as the system-provided version, but it seems non-trivial. Considering the amount of suid executables and other configuration involved in rootless containers, I think I will probably abandon this effort.

**Comment by bernhardkaindl** [comment]
> @alecbcs @bernhardkaindl I tried getting the spack-built buildah to work as well as the system-provided version, but it seems non-trivial. Considering the amount of suid executables and other configuration involved in rootless containers, I think I will probably abandon this effort.

For rootless containers, if they don't need to have the `docker` ecosystem, you can use Sigularity / Apptainer:

https://apptainer.org/
https://github.com/apptainer/apptainer
https://docs.sylabs.io/guides/3.5/user-guide/introduction.html
https://en.wikipedia.org/wiki/Singularity_(software)

Both variants of it have recipes in spack.

The build of singularity least once worked for me rootless when I worked on it and made some fixes.

Could you please close this PR if your decision to abadon getting buildah working for rootless containers?

**Comment by upsj** [comment]
I tried spack-built apptainer variants before, but they didn't work out of the box, so I went with a system-installed version instead.

**Pull request closed without merging**
